src/functional/fencil_processors/source_modules/cpp_gen.py

```diff
@@ -14,15 +14,35 @@
 
 
 import ctypes
+from dataclasses import dataclass
 from types import MappingProxyType
 from typing import Final, Sequence, Type
 
 import numpy
 
-from functional.fencil_processors import source_modules
+from eve.codegen import format_source
+from functional.fencil_processors.source_modules import source_modules
 
 
-LANGUAGE_ID: Final = "cpp"
+@dataclass(frozen=True)
+class CppLanguage:
+    """Implements source_modules.IncludeImplementationLanguageProtocol for C++ language settings."""
+
+    name: str
+    implementation_extension: str
+    include_extension: str
+    formatting_style: str
+
+    def format_source(self, source_code: str) -> str:
+        return format_source(self.name, source_code, style=self.formatting_style)
+
+
+CPP_DEFAULT: Final = CppLanguage(
+    name="cpp",
+    implementation_extension=".cpp",
+    include_extension=".cpp.inc",
```

Comment:
I see it like this: if someone writes a backend that separates headers cleanly, then they are free to provide a different default `CppLanguage` instance, which sets `implementation_extension` to "hpp".

Only once someone really writes a backend that needs three estensions at the same time should there be a new language protocol and a default implementation for it. Seems pointless otherwise.

Reply:
What use case have you had in mind for this extensions feature? I don't really understand how it's connected to unique names.

Include guards don't help here, they prevent multiple inclusions into the same translation unit, but they allow inclusion into many translation units. The problem with `.cpp.inc` is that it should be included into no more than one translation unit.